airflow/models.py

```diff
@@ -1288,19 +1288,19 @@ def run(
             verbose=True)
 
         if not runnable and not mark_success:
-            if self.state != State.QUEUED:
-                # If a task's dependencies are met but it can't be run yet then queue it
-                # instead
-                self.state = State.QUEUED
-                msg = "Queuing attempt {attempt} of {total}".format(
-                    attempt=self.try_number % (task.retries + 1) + 1,
-                    total=task.retries + 1)
-                logging.info(hr + msg + hr)
-
-                self.queued_dttm = datetime.now()
-                msg = "Queuing into pool {}".format(self.pool)
-                logging.info(msg)
-                session.merge(self)
+            # we might have hit concurrency limits, which means we probably
+            # have been running prematurely.
+            self.state = State.UP_FOR_RETRY
```

Comment:
Even if this is included it's still possible for the DAG to run out of execution slots I believe in the current implementation (the scheduler doesn't keep a count of the tasks it has sent of for execution but haven't run yet).

I think the reason we didn't include DagTISlotsAvailable in QUEUE_DEPS that the scheduler uses (which is the only dependency in RUN_DEPS that is not in QUEUE_DEPS) is because we wanted to queue up and show as grey task instances that weren't running due to concurrency limits (to preserve the old behavior). I feel like this isn't important so maybe we can just replace QUEUE_DEPS with RUN_DEPS in the scheduler.